### PR TITLE
PCHR-1881: Fix Effective date validation not done when contract revis…

### PR DESCRIPTION
**Pull request by bmnepali**
…ion is edited from Full History

### Issue:
While editing a revision form the list of available revisions form Full History Tab in Contract revisions, the effective date is not validated with existing revisions having same effective date. As a result, different revisions can have same effective date.

### Solution:
1. When Saving the revision, validate the effective date before  saving the revision data using `ContractRevisionService.validateEffectiveDate({})`.
2. Removing the a `ng-click="save()"` in "Save and Done" button from`modalChangeReason.html` as save is already called while submitting the form. (Removes call to save() twice)

### Screenshots:
1. Click to edit a revision form Full History tab form Contract: Dislays the Reason Change Modal.
screen shot 2017-01-23 at 2 24 32 pm

2. In effective date field, choose a effective date that is already taken by other revisions: It is display error message and allows user to change the effective date without closing the Reason Change Modal.
screen shot 2017-01-23 at 2 25 00 pm

### CODE REVIEW

File: hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
-    'job-contract/controllers/controllers'
+    'job-contract/controllers/controllers',
+    'job-contract/services/contract'
 ], function (moment, controllers) {
     'use strict';
 
-    controllers.controller('ModalChangeReasonCtrl',['$scope','$uibModalInstance', 'content', 'date', 'reasonId', '$log',
-        function ($scope, $modalInstance, content, date, reasonId, $log) {
+    controllers.controller('ModalChangeReasonCtrl',['$scope','$uibModalInstance', 'content', 'date', 'reasonId', '$log', 'settings', 'ContractRevisionService',
+        function ($scope, $modalInstance, content, date, reasonId, $log, settings, ContractRevisionService) {
```
Comment (dhavalsharma) [inline hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js]: group similar together like $log must be along with other system dependencies like $scope

File: hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js [hunk 2]
```diff
@@ -26,9 +27,20 @@ define([
             }
 
             $scope.save = function () {
-                $modalInstance.close({
-                    reasonId: $scope.change_reason,
-                    date: $scope.effective_date ? moment($scope.effective_date).format('YYYY-MM-DD') : ''
+                ContractRevisionService.validateEffectiveDate({
+                    contact_id: settings.contactId,
```
Comment (dhavalsharma) [inline hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js]: 4 spaces

File: hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js [hunk 3]
```diff
@@ -26,9 +27,20 @@ define([
...
+                    effective_date: $scope.effective_date
+                }).then(function(result){
```
Comment (dhavalsharma) [inline hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js]: better to have .then on new line like 
```
})
.then
```

File: hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js [hunk 4]
```diff
@@ -26,9 +27,20 @@ define([
...
+                    if (result.success) {
+                        $modalInstance.close({
+                          reasonId: $scope.change_reason,
+                          date: $scope.effective_date ? moment($scope.effective_date).format('YYYY-MM-DD') : ''
+                        });
+                    } else {
+                        CRM.alert(result.message, 'Error', 'error');
+                        $scope.$broadcast('hrjc-loader-hide');
+                    }
+                },function(reason){
```
Comment (dhavalsharma) [inline hrjobcontract/js/src/job-contract/controllers/modal/modal-change-reason.js]: ps remove handler if `reason` is not used

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
+define([
+    'common/moment',
+    'common/angularMocks',
+    'job-contract/app'
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: i think we are doing controller testing here so ps include the controller only

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+], function(moment) {
+  'use strict';
+
+  describe('ModalChangeReasonCtrl', function() {
+    var $q, $rootScope, $scope, $controller, ctrl, modalInstanceSpy, ContractRevisionServiceSpy;
+
+    beforeEach(module('hrjc', 'job-contract.templates'));
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: i think we are doing controller testing here so ps include the controller module only

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      $controller = _$controller_;
+      $q = _$q_;
+      $rootScope = _$rootScope_;
+      ContractRevisionServiceSpy = ContractRevisionService;
+      modalInstanceSpy = jasmine.createSpyObj('modalInstanceSpy', ['dismiss', 'close']);
+
+      makeController();
+    }));
+
+    describe("Check if ", function() {
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: check if is incomplete instead try something like "Has all methods"

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      it(" save() and cancel() fuctions are defined", function() {
+        expect($scope.save).toBeDefined();
+        expect($scope.cancel).toBeDefined();
+      });
+    });
+
+    beforeEach(function() {
+      spyOn(window.CRM, 'alert');
+    });
+
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: move this beforeeach to main describe or rather include it in inject beforeeach

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    describe("Check if ", function() {
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: use english like sentence like "When saving change reason form"

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            success: false,
+            message: 'Sample alert message'
+          });
+
+          return deferred.promise;
+        });
+      });
+
+      it("ValidateEffectiveDate is called on saving Change Reason Form ", function() {
+        $scope.save();
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: can move to before each

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        expect(ContractRevisionServiceSpy.validateEffectiveDate).toHaveBeenCalled();
+      });
+
+      it("Modal's close function is called on effective_date validation success ", function() {
+        $scope.save();
+        $scope.$digest();
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: below can move to before each
```
$scope.save();
$scope.$digest();
```

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js (hunk 5 above)
Comment (bmnepali) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: We need `app.js` as we need `ContractRevisionService` for effective_date validation.

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js (hunk 6 above)
Comment (bmnepali) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: We need `hrjc` as we need `ContractRevisionService` for effective_date validation. Which will not be available on `hrjc.controllers` module.

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    }));
+
+    describe("when saving change reason form ", function() {
+      it(" should have save() and cancel() fuctions defined", function() {
+        expect($scope.save).toBeDefined();
+        expect($scope.cancel).toBeDefined();
+      });
+    });
+
+    describe("when saving change reason form ", function() {
```
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: combine above 2 describe as they are same

File: hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js (hunk 5 above)
Comment (dhavalsharma) [inline hrjobcontract/js/test/controllers/modal/modal-change-reason_test.js]: ps fix the error as that would be more appropriate testing

**Pull request merged**